### PR TITLE
Fix for WAM bug on Windows Server 2016 and Adding Support for Windows Server 2022

### LINKED ISSUE #3040: [Bug] Support Windows 2022 for WAM
**Is your feature request related to a problem? Please describe.**
WAM Broker probably fails to detect Windows 2022 as a supported OS.

**Describe the solution you'd like**
Support Windows Server 2022

**Additional context**
New OS versions were released that need to be added to DesktopOsHelper.
ProductName from Registry:   "Windows Server 2022 Datacenter"

### PR DESCRIPTION
**Pull request by gladjohn**
Fixes #3040 #2946 

**Changes proposed in this request**
- Adding WAM support for new Server OS (Windows Server 2022)
- WAM is not available on Win Server 2016 so removing support and falling back to browser

**Testing**
manual testing 

**Performance impact**
none

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs [hunk 1]
```diff
@@ -82,35 +80,53 @@ public static bool IsMac()
...
+        /// So we are just checking for the "Windows 10" version string. This is by design 
+        /// https://microsoft.visualstudio.com/OS/_workitems/edit/36426651
+        /// </summary>
+        /// <returns>Returns <c>true</c> if the Windows Client Version has WAM support</returns>
+        private static bool IsWin10OrNewerInternal()
         {
-            return s_win10Lazy.Value;
+            string winVersion = GetWindowsVersionString();
+
+            if (winVersion.Contains("Windows 10", StringComparison.OrdinalIgnoreCase))
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs:94]: Is the value same across all cultures and languages?

**Review by SameerK-MSFT — APPROVED** [review]

**Comment by SameerK-MSFT** [comment]
I think this will require manual test in the next release. (And not again till a new Windows OS shows up)

File: src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs [hunk 2]
```diff
@@ -82,35 +80,53 @@ public static bool IsMac()
...
-                {
-                    return true;
-                }
+            if (winVersion.Contains("Windows Server 2019", StringComparison.OrdinalIgnoreCase) ||
+            winVersion.Contains("Windows Server 2022", StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
             }
 
             return false;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs:116]: Can we go a step further and try to extract (e.g. using a regexp) the year (i.e. if this starts with Windows Server) ? We could then compare it with 2019.

File: src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs [hunk 3]
```diff
@@ -82,35 +80,53 @@ public static bool IsMac()
...
+
+            return false;
         }
 
-        private static bool IsWin10Internal()
+        /// <summary>
+        /// Checks the retrieved Windows version to see if it is a WAM supported Windows Server OS. 
+        /// </summary>
+        /// <returns>Returns <c>true</c> if the Windows Server Version has WAM support</returns>
+        private static bool IsWinServer2019OrNewerInternal()
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs:106]: Did you try the approach with `RtlGetVersion` Windows API from https://stackoverflow.com/questions/2819934/detect-windows-version-in-net  ? Just curious what it returns on win server.

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs:106]: Another way to look at this is - can we more reliably detect Win Server 2016 and Win 8 / Win 8.1 ? Anything else we can consider supported?

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
Need a better approach for future Win Server OS, as we cannot ask applications to upgrade MSAL anytime a new Win Server comes out. See some suggestions.

**Comment by bgavrilMS** [comment]
What was the problem with the builds?

**Comment by bgavrilMS** [comment]
Why were the builds failing for the rtf stuff?

**Comment by gladjohn** [comment]
> Why were the builds failing for the rtf stuff?

Build failed because of the usage "Error CS0618: 'Marshal.SizeOf(Type)' is obsolete", fixed that but found another issue with Windows 11. Windows 10/11 multi session ProductType returns 3 (NT_Server). Per, [docs](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.microsoft.com%2Fen-us%2Fwindows%2Fsecurity%2Fthreat-protection%2Fwindows-firewall%2Fcreate-wmi-filters-for-the-gpo&data=04%7C01%7CGladwin.Johnson%40microsoft.com%7C8e8e7972b97c4741df9908d9db84f466%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637782186350524291%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=0iPkjbuVNHXBN%2FAvbf7VVdzeQ4pSjiMwA0L3X9L12xE%3D&reserved=0) "For server operating systems that are not domain controllers and for Windows 10 and Windows 11 multi-session, use ProductType="3"." Building a non multi session Windows 11 to test

File: src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs (hunk 3 above)
Comment (gladjohn) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs]: Using RtlGetVersion to detect Window10 and Above and Server 2019 and Above for WAM support

File: src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs (hunk 2 above)
Comment (gladjohn) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs]: used and RtlGetVersion

File: src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs (hunk 1 above)
Comment (gladjohn) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DesktopOsHelper.cs]: Tested one locale and it is in English.

File: src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/Win32API.cs [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        }
+
+        /// <summary>
+        /// Checks if the OS supports WAM (Web Account Manager)
+        /// WAM Supported OS's are Windows 10 and above for Client, Windows 2019 and above for Server
+        /// </summary>
+        /// <returns>Returns <c>true</c> if the OS Version has WAM support</returns>
+        public static bool IsWamSupportedOs()
+        {
+            try
```
[the remark this answers is not in the document]
Comment (gladjohn) [reply src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/Win32API.cs]:
```cs
private static Lazy<bool> s_wamSupportedOSLazy = new Lazy<bool>(
           () => IsWamSuportedOSInternal());```

as discussed, it is in the desktoposhelper

**Review by bgavrilMS — APPROVED** [review]
LGTM

**Comment by gladjohn** [comment]
> I think this will require manual test in the next release. (And not again till a new Windows OS shows up)

added to test sheet

**Pull request merged**